src/Utilities/StdArrayHelpers.hpp

```diff
@@ -166,6 +175,64 @@ decltype(auto) magnitude(const std::array<T, 3>& a) noexcept {
 }
 //@}
 
+/// \ingroup UtilitiesGroup
+/// \brief Dot product between two arrays
+///
+/// \details This also works elementwise if T is a container and R is a float or
+/// the other way round. The return type will always be the iterable type. If
+/// both or none are iterable, the return type will be T, the first argument
+/// type.
+///
+/// \requires The return type must have operator+= implemented. Elementwise
+/// multiplication between T and R must be defined. In addition, the sizes of
+/// all containers within T and R must match. If T and R are both iterable, the
+/// size of each container in R must be greater or equal than the size of each
+/// container in T.
+
+// `DotReturnType_t` will always be T, unless R is iterable and T is not
+// iterable.
+template <typename T, typename R, typename = std::nullptr_t>
+struct DotReturnType {
+  using type = tt::remove_cvref_wrap_t<T>;
+};
+
+template <typename T, typename R>
+struct DotReturnType<
+    T, R, Requires<not tt::is_iterable_v<T> and tt::is_iterable_v<R>>> {
+  using type = tt::remove_cvref_wrap_t<R>;
+};
+
+template <typename T, typename R>
+using DotReturnType_t = typename DotReturnType<T, R>::type;
+
+template <size_t Dim, typename T, typename R>
+void dot(const gsl::not_null<tt::remove_cvref_wrap_t<T>*> res,
+         const std::array<T, Dim>& first, const std::array<R, Dim>& second) {
+  *res = 0.;
+  auto second_it = second.begin();
+  for (auto first_it = first.begin(); first_it != first.end();
+       ++first_it, ++second_it) {
+    *res += *first_it * *second_it;
```

Comment:
this is nice and general, though what do you think about just implementing `dot` like `magnitude` above for 1, 2 and 3 dimensions? The simplicity of that implementation is hard to beat :) We probably don't need a `dot` function for higher dimensions.

Reply:
hmm, yes the body is nice but the signature is rather horrible 
```
void dot(
    const gsl::not_null<tt::remove_cvref_wrap_t<detail::DotReturnType_t<T, R>>*>
    res,
    const std::array<T, 3>& first, const std::array<R, 3>& second) 
```
that the whole function isnt too easy to read and Im not sure I want to repeat that 3 times. I dont have a strong opinion.

Reply:
How about this:

```cpp
template <typename R, typename T, typename U>
void dot(
    const gsl::not_null<std::array<R, 3>*> result,
    const std::array<T, 3>& first, const std::array<U, 3>& second) 
```

Reply:
Im not sure I prefer it. Its slighly nicer to read but there is zero type checking done on `R`. The user can pass in any object that has double assignment and `+=` implemented... this is great if the user wants to do hacky things but not great if template substitution succeeds where it should fail. Probably not a big issue though.

Reply:
could you add a `static_assert(std::is_same_v<R, decltype(T * U)>);` (probably needs some adjust to handle convertible things, references, etc.

Reply:
this is a good point. Is it possible to get rid of these entire `DotReturnType` structs by inferring the return type from `T * R`? 

I was trying something like `decltype(T() * R())` which seems to compile to my surprise but the blaze expression templates are ruining the fun since they dont appear to have constructors. Can these be "decayed" down somehow to `DataVector`s?

Reply:
Yep, this is what is often done in generic libraries. You should be able to use the type alias `get_vector_element_type_t` defined in `VectorImpl.hpp` to avoid the expression templates

Reply:
unfortunately `VectorImpl.hpp` already includes `StdArrayHelpers.hpp` and untangling that cyclic dependency seems like a mess